[clang][analyzer][doc] Migrate user-related docs from HTML to RST

gamesh411 opened the pull request:
User documentation currently found at https://clang-analyzer.llvm.org is migrated to RST format.

This commit migrates all the relevant content, including parts that are suspicious or even clearly outdated. These issues will be cleaned up in separate follow-up commits (where the diff is not obscured by the format change). However, a few typos are fixed and some parts (availability of binary releases, integration with XCode) are marked (with [Legacy] tags) to highlight that they are outdated. 

The primary motivation for this change is to update the facts in the docs and make them discoverable from the RST-generated doc-tree as well (many subpages are not accessible **at all**, as the menu generation for the HTML based page is not working at all).

gamesh411 committed:
[clang][analyzer][doc] Migrate user-related docs from HTML to RST

User-documentation now found at https://clang-analyzer.llvm.org is
migrated to RST format.
The content of the documentation is updated to reflect recent
availability of binary releases and integration with XCode and
commandline tools.

gamesh411 committed:
merge scan-build and codechecker docs into command-line-usage

gamesh411 committed:
add filing bugs doc

gamesh411 committed:
update xcode doc links

gamesh411 committed:
update installation docs

gamesh411 committed:
inline legacy latest build

gamesh411 commented:
Some parts, especially Apple-related sections, may no longer be relevant in the docs.
Please share insights into which parts should be corrected or omitted from the migration process.
The legacy Clang builds (the most recent one uses Clang version 4.0.0) are still on the clang-analyzer.llvm.org site, and this seems to be the only part that requires the HTML page and the hosting provider (AFAIK).

steakhal commented on clang/docs/analyzer/user-docs/Installation.rst:
```diff
@@ -0,0 +1,37 @@
+Obtaining the Static Analyzer
+=============================
+
+This page describes how to download and install the analyzer. Once the analyzer is installed, follow the asdf :doc:`CommandLineUsage` on using the commandline to get started analyzing your code.
```
`asdf` appears to be a typo.

NagyDonat replied:
Nice catch! :laughing:

steakhal commented on clang/docs/analyzer/user-docs/CommandLineUsage.rst:
```diff
@@ -0,0 +1,238 @@
+Static Analyzer is by design a GUI tool originally intended to be consumed by the XCode IDE.
+Its purpose is to find buggy execution paths in the program, and such paths are very hard to comprehend by looking at a non-interactive standard output.
+It is possible, however, to invoke the Static Analyzer from the command line in order to obtain analysis results, and then later view them interactively in a graphical interface.
+The following tools are used commonly to run the analyzer from the commandline.
```
Interestingly, there are a few variations for spelling `command line`, `commandline`, `command-line`.
I'd appreciate if we could harmonize these :D

steakhal reviewed (changes requested):
I mostly had some higher level issues. Thanks for updating the docs.

gamesh411 commented on clang/docs/analyzer/user-docs/CommandLineUsage.rst:
```diff
@@ -0,0 +1,238 @@
+    - Perfect for managing large amounts of Static Analyzer warnings in a collaborative environment.
+    - Generally much more feature-rich than scan-build.
+    - Supports incremental analysis: Results can be stored in a database, subsequent analysis runs can be compared to list the newly added defects.
+    - :doc:`CrossTranslationUnit` is supported fully on Linux via CodeChecker.
```
I have checked, and I think the GitHub viewer is buggy.
Steps used to verify:
```
// inside the build directory
1. build target `docs-clang-html` (cmake --build . -- docs-clang-html)
2. start webserver (python3 -mhttp.server -d tools/clang/docs/html)
3. browse localhost
```

steakhal replied:
Sounds good to me. I don't really wanna put the effort into checking it out.

gamesh411 commented on clang/docs/analyzer/user-docs/CommandLineUsage.rst:
```diff
@@ -0,0 +1,238 @@
+The following tools are used commonly to run the analyzer from the commandline.
+Both tools are wrapper scripts to drive the analysis and the underlying invocations of the Clang compiler:
+
+1. CodeChecker_ is a driver and web server that runs the Static Analyzer on your projects on demand and maintains a database of issues.
```
As above, this is not the case for the actual HTML generated.

NagyDonat replied:
It's still strange that there is only one underscore which is _after_ the word 'CodeChecker'. I'd guess that the intent was to italicize 'CodeChecker' and we should add another underscore _before_ this word for the sake of clarity (even if the actual HTML generation can work with this half-broken situation.

steakhal replied:
Yes, please have a look.

gamesh411 replied:
The intent was to create a link to the subheading lower down in the document, not to italicize it.

NagyDonat replied:
Oh, that's surprising -- I was not familiar with that feature of RST. Naturally this invalidates my concern.

gamesh411 committed:
remove typo

gamesh411 committed:
unify usages of 'command line'

gamesh411 committed:
switch the order CodeChecker and scan-build

steakhal commented:
I'm pretty much good now with this PR. I think this is a net positive.
Maybe consider redirection instead of deleting the pages.

NagyDonat reviewed (commented):
I'm really happy to see that this duplicated documentation is finally cleaned up after so many years :partying_face:

I added a few minor remarks in inline comments, but the change looks good overall.

As we discussed privately, this change should be (mostly) limited to the format changes, and then separate follow-up commit should eliminate the outdated content (and, perhaps, add new information if that's necessary). I'll update the PR description to emphasize this.

haoNoQ commented on clang/www/analyzer/index.html:
```diff
@@ -96,11 +96,11 @@ <h2>Download</h2>
    <ul>
     <li>Latest build (10.8+):<br>
-     <!--#include virtual="latest_checker.html.incl"-->
+     <b><a href="downloads/checker-279.tar.bz2">checker-279.tar.bz2</a></b> (built November 14, 2016)
```
The entire downloads section should probably be nuked. The macOS section is completely out of date and meaningless. The rest boils down to "just download clang how you normally download clang".

Instead, we can point them to the scan-build and CodeChecker page, or to their IDE's documentation.

I suggest:
```txt
The static analyzer is part of Clang. Please refer to the
<a href="https://releases.llvm.org/download.html">LLVM
releases page</a> for further instructions.

The official LLVM releases also include <a href="scan-build.html">scan-build</a>.
If you're installing Clang from a different source, such as
a linux package repository, then scan-build may be packaged
separately as an individual package, or as part of
a "clang tools" package.

If your IDE is using Clang, it may natively integrate the static analyzer.
On macOS, the easiest way to use the static analyzer is to invoke it
<a href="https://clang.llvm.org/docs/analyzer/user-docs/UsingWithXcode.html">directly
from Xcode</a>.

Additionally, if you're using <a href="https://clang.llvm.org/extra/clang-tidy/">clang-tidy</a>,
you can naturally make the static analyzer run alongside clang-tidy
by enabling the <a href="https://clang.llvm.org/extra/clang-tidy/checks/list.html">clang-analyzer</a>
checks.
```

(Could be a follow-up commit.)

gamesh411 replied:
@haoNoQ Thanks for the insights!
I'll also create another PR to clean up the index.html and finish the developer section.
There will be questions about what to do with open projects and checker ideas, but let's discuss it later in that PR.

gamesh411 committed:
use consistent spelling of 'the static analyzer'

gamesh411 committed:
add redirect pages

gamesh411 commented:
* Updated the HTML pages to use the redirect logic in available_checks.html (instead of deleting them).
* Updated the usages of 'the static analyzer' in the files touched by this patch.
@steakhal, @NagyDonat, @haoNoQ Thanks for the snappy review!
